docs: add some missing features to overview pages

silvenon opened the pull request:
We want to always have breadcrumbs, so we're adding a reasonable default when they aren't explicitly provided.

 Storybook: https://orbit-docs-default-breadcrumbs.surge.sh

CollierCZ reviewed (dismissed):
Would it be possible to quickly implement something for multiple levels (like https://e51deade.orbit.pages.dev/components/overlay)? If not quickly, I don't think it's worth changing, so this is fine. Just if it's not too long, it'd be nice to have.

silvenon commented:
Should be good now. 👍 As a bonus I added trailing slashes to overview pages and fixed that the necessary navigation items open. I refactored the code for overview pages and now it's being tested better.

silvenon commented on docs/services/overview-pages.js:
```diff
@@ -51,39 +55,55 @@ function createPages(obj, slug) {
+  );
 
-  files.forEach(file => {
-    const wholePath = omitNumbers(file)
-      .replace(/\.[^/.]+$/, "")
-      .split("/");
```
This was unclear, so I tried to clear it up with `path.parse`.

silvenon commented on docs/services/overview-pages.js:
```diff
@@ -51,39 +55,55 @@ function createPages(obj, slug) {
-    callback(page);
-    return renderOverviewPages(pages, callback);
-  });
+const forEachNestedPage = (pages, callback) => {
```
This function is very similar, but I renamed it for clarity.

silvenon commented on docs/services/overview-pages.js:
```diff
@@ -51,39 +55,55 @@ function createPages(obj, slug) {
 };
 
-async function createOverviewPages(callback) {
+async function getOverviewPages() {
```
I renamed this function so that it better describes that it now returns an array of overview pages.

silvenon commented on docs/utils/document.js:
```diff
@@ -1,6 +1,6 @@
 const path = require("path");
 
-const rePathNumbers = new RegExp(`(${path.sep})\\d+-`, "g");
+const rePathNumbers = new RegExp(`(^|${path.sep})\\d+-`, "g");
```
This was needed so that `omitNumbers` works for `01-*` as well, although I'm not sure if it's needed now, but it doesn't hurt.

CollierCZ replied:
I don't see any cases where omitNumbers is passed a relative path where it might be an issue, but I don't suppose it hurts.

While looking it up, i did notice we now have two `omitNumbers` functions. I'd get rid of the one in `docs/src/utils` since it doesn't seem to be used anywhere. Nor is `capitalize` in that file, though it doesn't have another function it might get confused with.

silvenon replied:
In the test environment we use relative paths, see in [`gatsby-node.js`](https://github.com/kiwicom/orbit/pull/3005/files#diff-97ce157ffa078c33fbccfb2e69ae8be278c4074fbd0ec7c750ce35cceaa454f1R193).

silvenon commented on docs/src/components/DocNavigation/index.tsx:
```diff
@@ -39,7 +39,7 @@ export function groupTrails(trails: Trail[]): Navigation {
         acc[name] = { result: [] };
         if (index < trail.length - 1) {
-          acc.result.push({ type: "branch", name, items: acc[name].result });
+          acc.result.push({ type: "branch", name, url, items: acc[name].result });
```
This was needed to determine which navigation item to expand.

silvenon reviewed (commented):
A little help with reviewing.

CollierCZ commented on docs/services/overview-pages.js:
```diff
@@ -51,39 +55,55 @@ function createPages(obj, slug) {
+  // this way we're essentially flattening the pages to be only one level deep
+  forEachNestedPage(allPages, page => {
+    overviewPages.push(page);
   });
```
As above, I'm not sure this logic is necessary before the page is created, but if it's here, wouldn't it be a good use for reduce?

silvenon replied:
I made this logic slightly more clear by inlining the recursive function at the place of invocation. I don't think this would be a good use case for `reduce` because `reduce` is best for deriving a single non-array value out of an array of elements. It's being overused in JavaScript because you can chain it to other array methods, but it often results in very unclear logic.

CollierCZ reviewed (changes requested):
A couple of minor things probably not fully related to this PR. It'd be nice to at least clean up the `omitNumbers` function, though. Otherwise, it looks good to me.

silvenon commented on docs/gatsby-node.js:
```diff
@@ -180,13 +182,19 @@ exports.createPages = async ({ graphql, actions, reporter }) => {
+        path: page.slug,
+        component:
+          process.env.NODE_ENV === "test" ? path.relative(ROOT, overviewPath) : overviewPath,
+        context: { ...page, trail },
```
In order to programmatically create pages in Gatsby a lot of this stuff had to be computed in advance. We can probably fetch some of this data via GraphQL, which I think would split the logic somewhat, and we will because we need it for other stuff, I'd postpone this as a separate effort.
